File: src/web/routes/trimRoutes.ts
/**
 * Trim routes - Create and delete trimmed versions of videos
 */

import type { Router} from "express";
import { Database } from "../../core/database.js";
import { TrimService } from "../../core/worker/trimService.js";
import { MuxingError } from "../../types/errors.js";
import { asyncHandler } from "./errorHandler.js";
import { createTrimSchema, formatValidationErrors } from "../../types/schemas.js";
import { Logger } from "../../core/logger.js";

const logger = Logger.getInstance();

export function registerTrimRoutes(router: Router): void {
  // POST /api/jobs/:id/trims - Create trim
  router.post(
    "/jobs/:id/trims",
    asyncHandler(async (req, res) => {
      const jobId = req.params.id as string;

      // Validate request body with Zod schema
      const validation = createTrimSchema.safeParse(req.body);
      if (!validation.success) {
        const errors = formatValidationErrors(validation.error);
        return res.status(400).json({
          error: "Validation failed",
          details: errors
        });
      }

      const { startTime, endTime } = validation.data;

      // Load job
      const db = await Database.getInstance();
      const job = await db.getJob(jobId);
      if (!job) {
        return res.status(404).json({ error: "Job not found" });
      }

      // Create AbortController to cancel on client disconnect
      const abortController = new AbortController();
      req.on("close", () => {
        abortController.abort();
      });

      // Create trim with abort signal
      try {
        const trim = await TrimService.createTrim(
          job,
          startTime,
          endTime,
          abortController.signal,
        );
        res.json({ trim });
      } catch (error: unknown) {
        const message = error instanceof Error ? error.message : String(error);
        logger.error(`[TrimRoutes] Failed to create trim: ${message}`);
        const status = error instanceof MuxingError ? 500 : 400;
        res.status(status).json({ error: "Failed to create trim" });
      }
    }),
  );

  // DELETE /api/jobs/:id/trims/:trimId - Delete trim
  router.delete(
    "/jobs/:id/trims/:trimId",
    asyncHandler(async (req, res) => {
      const { id: jobId, trimId } = req.params;
      await TrimService.deleteTrim(jobId as string, trimId as string);
      res.json({ success: true });
    }),
  );

  // GET /api/jobs/:id/trims - List trims
  router.get(
    "/jobs/:id/trims",
    asyncHandler(async (req, res) => {
      const jobId = req.params.id as string;
      const db = await Database.getInstance();
      const job = await db.getJob(jobId);
      if (!job) {
        return res.status(404).json({ error: "Job not found" });
      }
      res.json({ trims: job.trims || [] });
    }),
  );
}

File: internal/web/routes/trim_routes.go
package routes

import (
	"encoding/json"
	"math"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/vampiricwulf/Moombox/internal/database"
	"github.com/vampiricwulf/Moombox/internal/worker"
)

// TrimRoutes registers trim-related API routes.
func TrimRoutes(r chi.Router, db *database.Database, trimSvc *worker.TrimService) {
	// POST /api/jobs/:id/trims
	r.Post("/api/jobs/{id}/trims", func(rw http.ResponseWriter, req *http.Request) {
		jobID := chi.URLParam(req, "id")
		job, err := db.GetJob(jobID)
		if err != nil || job == nil {
			jsonError(rw, "job not found", http.StatusNotFound)
			return
		}

		var body struct {
			StartTime *float64 `json:"startTime"`
			EndTime   *float64 `json:"endTime"`
		}
		if err := json.NewDecoder(req.Body).Decode(&body); err != nil {
			jsonError(rw, "invalid request body", http.StatusBadRequest)
			return
		}

		// Validate trim parameters (matching TypeScript createTrimSchema)
		if body.StartTime == nil || body.EndTime == nil {
			jsonError(rw, "Validation failed: startTime and endTime are required", http.StatusBadRequest)
			return
		}
		if math.IsNaN(*body.StartTime) || math.IsInf(*body.StartTime, 0) || math.IsNaN(*body.EndTime) || math.IsInf(*body.EndTime, 0) {
			jsonError(rw, "Validation failed: startTime and endTime must be finite numbers", http.StatusBadRequest)
			return
		}
		if *body.StartTime < 0 {
			jsonError(rw, "Validation failed: Start time cannot be negative", http.StatusBadRequest)
			return
		}
		if *body.EndTime <= *body.StartTime {
			jsonError(rw, "Validation failed: End time must be after start time", http.StatusBadRequest)
			return
		}
		if *body.EndTime-*body.StartTime < 1 {
			jsonError(rw, "Validation failed: Trim must be at least 1 second", http.StatusBadRequest)
			return
		}

		record, err := trimSvc.CreateTrim(req.Context(), job, *body.StartTime, *body.EndTime)
		if err != nil {
			// Match TS: don't expose internal error details
			jsonError(rw, "Failed to create trim", http.StatusBadRequest)
			return
		}

		jsonResponse(rw, map[string]any{"trim": record})
	})

	// DELETE /api/jobs/:id/trims/:trimId
	r.Delete("/api/jobs/{id}/trims/{trimId}", func(rw http.ResponseWriter, req *http.Request) {
		jobID := chi.URLParam(req, "id")
		trimID := chi.URLParam(req, "trimId")

		if err := trimSvc.DeleteTrim(jobID, trimID); err != nil {
			jsonError(rw, "failed to delete trim", http.StatusBadRequest)
			return
		}

		jsonResponse(rw, map[string]any{"success": true})
	})
}
